src/Microsoft.OData.Core/UriParser/Parsers/ODataPathParser.cs

```diff
@@ -1432,5 +1437,150 @@ private void CheckDollarEachSegmentRestrictions(int index)
+                if (entitySetSegment != null || navigationPropertySegment != null) // entitySet or  Navigation property
+                {
+                    IEdmVocabularyAnnotatable target;
+                    string kind, name;
+                    if (entitySetSegment != null)
+                    {
+                        target = entitySetSegment.EntitySet;
+                        kind = "entity set";
+                        name = entitySetSegment.EntitySet.Name;
+                    }
+                    else
+                    {
+                        target = navigationPropertySegment.NavigationProperty;
+                        kind = "navigation property";
+                        name = navigationPropertySegment.NavigationProperty.Name;
+                    }
+
+                    VerifyDerivedTypeConstraints(this.configuration.Model, target, fullTypeName, kind, name);
+                }
+                return;
+            }
+
+            // Navigation property: ~/Users(1)/Orders/NS.Cast
+            navigationPropertySegment = previous as NavigationPropertySegment;
+            if (navigationPropertySegment != null)
+            {
+                VerifyDerivedTypeConstraints(this.configuration.Model, navigationPropertySegment.NavigationProperty, fullTypeName, "navigation property", navigationPropertySegment.NavigationProperty.Name);
+                return;
+            }
+
+            // Structural property:  ~/Users(1)/Addresses/NS.Cast
+            PropertySegment propertySegment = previous as PropertySegment;
+            if (propertySegment != null)
+            {
+                IEdmProperty edmProperty = propertySegment.Property;
+                // Verify the DerivedTypeConstrictions on property.
+                VerifyDerivedTypeConstraints(this.configuration.Model, edmProperty, fullTypeName, "property", edmProperty.Name);
+
+                // Verify the Type Definition, the following codes should work if fix: https://github.com/OData/odata.net/issues/1326
+                /*
+                IEdmTypeReference propertyTypeReference = edmProperty.Type;
+                if (edmProperty.Type.IsCollection())
+                {
+                    propertyTypeReference = edmProperty.Type.AsCollection().ElementType();
+                }
+
+                if (propertyTypeReference.IsTypeDefinition())
+                {
+                    IEdmTypeDefinition edmTypeDefinition = propertyTypeReference.AsTypeDefinition().TypeDefinition();
+                    VerifyDerivedTypeConstraints(this.configuration.Model, edmTypeDefinition, fullTypeName, "type definition", edmTypeDefinition.FullName());
+                }
+                */
+
+                return;
+            }
+        }
+
+        private void CheckOperationTypeCastSegmentRestriction(IEdmOperation operation)
+        {
+            Debug.Assert(operation != null);
+
+            if (this.parsedSegments == null || !this.parsedSegments.Any())

```

Comment:
>|| !this.parsedSegments.Any() [](start = 44, length = 29)

This would also be caught in the this.parsedSegments.LastOrDefault() check on line 1546 below, right?

Reply:
Yes. You are right.

Reply:
# resolved.